### PR TITLE
linter: added a check for 'array' type hint

### PR DESCRIPTION
**Pull request by i582**
For `array` type hint for 

1. parameters of functions and methods;
2. return type of functions and methods;
3. class properties.

now necessary to add the more precise type to `phpdoc`.

For example:

```php
function f(array $a) {} // bad

/**
 * @param int[] $a
 */
function f1(array $a) {} // ok
```

For mixed types, need to write `mixed[]`.

```php
/**
 * @param mixed[] $a
 */
function f1(array $a) {} // ok
```

> Closures are not checked and are skipped.

### CODE REVIEW

File: src/linter/report.go [hunk 1]
```diff
@@ -650,6 +650,18 @@ $_ = is_real($a);`,
 			After: `$a = (float)100;
 $_ = is_float($a);`,
 		},
+
+		{
+			Name:     "typeHint",
+			Default:  false,
```
Comment (i582) [inline src/linter/report.go:656]: It seems to me that this check is too pedantic to be enabled by default.

File: src/linter/root.go [hunk 2]
```diff
@@ -1456,6 +1464,20 @@ func (d *rootWalker) parseFuncArgs(params []ir.Node, parTypes phpDocParamsMap, s
 	return args, minArgs
 }
 
+func (d *rootWalker) typeHintHasMoreAccurateType(typeHintType, phpDocType types.Map) bool {
+	// if is not array typehint
```
Comment (quasilyte) [inline src/linter/root.go:1468]: We usually write comments differently in noverify.
With periods and sentences start with a capital letter.

File: src/linter/root.go [hunk 3]
```diff
@@ -950,11 +945,17 @@ func (d *rootWalker) enterClassMethod(meth *ir.ClassMethodStmt) bool {
...
 	d.reportPhpdocErrors(meth.MethodName, doc.errs)
-	phpdocReturnType := doc.returnType
+	phpDocReturnType := doc.returnType
 	phpDocParamTypes := doc.types
 
 	class := d.getClass()
-	params, minParamsCnt := d.parseFuncArgs(meth.Params, phpDocParamTypes, sc, nil)
+
+	hintReturnType, has := d.parseTypeNode(meth.ReturnType)
+	if has && !doc.inherit {
```
Comment (quasilyte) [inline src/linter/root.go:954]: `has` is not a good variable name.
99% of time it should be just `ok` or, if you don't want to pollute the scope:
```go
if hintReturnType, ok := d.parseTypeNode(meth.ReturnType); ok {
  // Use hintReturnType
}
```

File: src/linter/root.go (hunk 2 above)
Comment (i582) [reply src/linter/root.go]: Done.

File: src/linter/root.go [hunk 4]
```diff
@@ -1413,40 +1416,45 @@ func (d *rootWalker) parseFuncArgs(params []ir.Node, parTypes phpDocParamsMap, s
...
 
-		if p.DefaultValue == nil && !parTyp.optional && !p.Variadic {
+		if p.DefaultValue == nil && !phpDocType.optional && !p.Variadic {
 			minArgs++
 		}
 
 		if p.VariableType != nil {
-			if varTyp, ok := d.parseTypeNode(p.VariableType); ok {
-				typ = varTyp
+			typeHintType, has := d.parseTypeNode(p.VariableType)
```
Comment (quasilyte) [inline src/linter/root.go:1432]: `has` is not a good variable name.

File: src/linter/root.go [hunk 5]
```diff
@@ -1413,40 +1416,45 @@ func (d *rootWalker) parseFuncArgs(params []ir.Node, parTypes phpDocParamsMap, s
...
+			if has {
+				if !isClosure && !d.typeHintHasMoreAccurateType(typeHintType, phpDocType.typ) {
+					d.Report(p, LevelNotice, "typeHint", "specify the type for the parameter $%s in phpdoc, 'array' type hint is not precise enough", p.Variable.Name)
```
Comment (quasilyte) [inline src/linter/root.go:1435]: I know it's easier to code it this way, but please do not `Report()` from parsing-related functions.
We should separate processing and linting as much as we can.
It makes refactoring impossible when `Report()` could be located anywhere.
Let's not make the situation worse.

File: src/linter/root.go [hunk 6]
```diff
@@ -1501,19 +1523,14 @@ func (d *rootWalker) enterFunction(fun *ir.FunctionStmt) bool {
...
-
 	d.checkCommentMisspellings(fun.FunctionName, fun.Doc.Raw)
 	d.checkIdentMisspellings(fun.FunctionName)
 	for _, p := range fun.Params {
 		d.checkVarnameMisspellings(p, p.(*ir.Parameter).Variable.Name)
 	}
 	doc := d.parsePHPDoc(fun.FunctionName, fun.Doc, fun.Params)
 	d.reportPhpdocErrors(fun.FunctionName, doc.errs)
-	phpdocReturnType := doc.returnType
+	phpDocReturnType := doc.returnType
```
Comment (quasilyte) [inline src/linter/root.go:1533]: I like the idea of making names consistent, but that could be done in a separate PR, with a reference to #422.
It's OK to keep these changes in this PR, but I'm almost sure there are more cases like this.
Extra diff lines make the review longer.

File: src/linter/root.go [hunk 7]
```diff
@@ -1522,13 +1539,18 @@ func (d *rootWalker) enterFunction(fun *ir.FunctionStmt) bool {
 
 	sc := meta.NewScope()
 
-	params, minParamsCnt := d.parseFuncArgs(fun.Params, phpDocParamTypes, sc, nil)
+	hintReturnType, has := d.parseTypeNode(fun.ReturnType)
```
Comment (quasilyte) [inline src/linter/root.go]: s/hintReturnType/returnTypeHint/?

File: src/tests/golden/testdata/embeddedrules/callSimplify.php [hunk 8]
```diff
@@ -1,5 +1,8 @@
 <?php
 
+/**
+ * @param  mixed[]  $array
```
Comment (quasilyte) [inline src/tests/golden/testdata/embeddedrules/callSimplify.php:4]: nit: there is an extra space between `mixed[]` and `$array`.

File: src/tests/golden/testdata/embeddedrules/offBy1.php [hunk 9]
```diff
@@ -1,5 +1,9 @@
 <?php
 
+/**
+ * @param  mixed[]  $xs
+ * @param  mixed[]  $tabs
```
Comment (quasilyte) [inline src/tests/golden/testdata/embeddedrules/offBy1.php:5]: I'm not sure why 2 spaces are used so consistently. Copy/paste maybe?

Comment (i582) [reply src/tests/golden/testdata/embeddedrules/offBy1.php:5]: Looks like IntelliJ auto format.

File: src/linter/root.go [hunk 10]
```diff
@@ -820,23 +820,23 @@ func (d *rootWalker) enterPropertyList(pl *ir.PropertyListStmt) bool {
...
-		specifiedType = typ
-	}
-
 	d.checkCommentMisspellings(pl, pl.Doc.Raw)
-	typ := d.parsePHPDocVar(pl, pl.Doc)
+	phpDocType := d.parsePHPDocVar(pl, pl.Doc)
+
+	typeHintType, has := d.parseTypeNode(pl.Type)
+	if has && !d.typeHintHasMoreAccurateType(typeHintType, phpDocType) {
+		d.Report(pl, LevelNotice, "typeHint", "specify the type for the property in phpdoc, 'array' type hint is not precise enough")
```
Comment (quasilyte) [inline src/linter/root.go:828]: I think this is more concise:
```
s/is not precise enough/too generic/`
```

File: src/linter/root.go (hunk 3 above)
Comment (i582) [reply src/linter/root.go]: Done.

File: src/linter/report.go (hunk 1 above)
Comment (quasilyte) [reply src/linter/report.go:656]: Agreed.

File: src/linter/root.go (hunk 4 above)
Comment (i582) [reply src/linter/root.go]: Done.

File: src/phpdoc/parser.go [hunk 11]
```diff
@@ -9,8 +9,9 @@ import (
 )
 
 type Comment struct {
-	Raw    string
-	Parsed []CommentPart
+	Raw     string
+	Parsed  []CommentPart
+	Inherit bool
```
Comment (quasilyte) [inline src/phpdoc/parser.go:14]: Maybe add an in-line comment that it means that there was `inheritdoc` somewhere in the parsed comment?
```
Inherit bool // comment contains @inheritdoc marker
```

File: src/linter/root.go (hunk 7 above)
Comment (i582) [reply src/linter/root.go]: Done.

File: src/tests/golden/testdata/embeddedrules/offBy1.php (hunk 9 above)
Comment (i582) [reply src/tests/golden/testdata/embeddedrules/offBy1.php]: Done.

File: src/phpdoc/parser.go (hunk 11 above)
Comment (i582) [reply src/phpdoc/parser.go]: Done.

File: src/tests/golden/testdata/embeddedrules/callSimplify.php (hunk 8 above)
Comment (i582) [reply src/tests/golden/testdata/embeddedrules/callSimplify.php]: Done.

File: src/linter/root.go (hunk 10 above)
Comment (i582) [reply src/linter/root.go]: Done.

File: src/linter/root.go (hunk 5 above)
Comment (i582) [reply src/linter/root.go]: Done.

**Pull request merged**